File: composeApp/src/jvmMain/java/app/typist/Correlation.java
package app.typist;

import java.util.ArrayList;
import java.util.stream.IntStream;

import condition.Condition;
import condition.ValueHolder;

/**
 * A class for calculating the correlation coefficient.
 */
public class Correlation extends Object
{

    /**
     * Holds the collection of X values (pixel values of a character image).
     */
    private ArrayList<Double> collectionX;

    /**
     * Holds the collection of Y values (pixel values of one block from the divided source image).
     */
    private ArrayList<Double> collectionY;

    /**
     * A flag to indicate whether the denominator is n-1.
     */
    private Boolean nMinusOne = true;

    /**
     * Holds the mean of X.
     */
    private Double meanX;

    /**
     * Holds the mean of Y.
     */
    private Double meanY;

    /**
     * Holds the variance of X.
     */
    private Double varianceX;

    /**
     * Holds the variance of Y.
     */
    private Double varianceY;

    /**
     * Holds the standard deviation of X.
     */
    private Double standardDeviationX;

    /**
     * Holds the standard deviation of Y.
     */
    private Double standardDeviationY;

    /**
     * Holds the covariance.
     */
    private Double covariance;

    /**
     * Holds the correlation coefficient.
     */
    private Double coefficient;

    /**
     * Initializes with empty X and Y collections.
     */
    public Correlation()
    {
        this.collectionX = new ArrayList<>();
        this.collectionY = new ArrayList<>();
    }

    /**
     * Initializes when only the X values are provided.
     *
     * @param xCollection The collection of pixel values from the character image.
     */
    public Correlation(ArrayList<Double> xCollection)
    {
        this.collectionX = xCollection;
        this.collectionY = new ArrayList<>();
    }

    /**
     * Initializes when both X and Y values are provided.
     *
     * @param xCollection The collection of pixel values from the character image.
     * @param yCollection The collection of pixel values from one block of the divided source image.
     */
    public Correlation(ArrayList<Double> xCollection, ArrayList<Double> yCollection)
    {
        this.collectionX = xCollection;
        this.collectionY = yCollection;
    }

    /**
     * Returns the correlation coefficient.
     *
     * @return The correlation coefficient.
     */
    public Double coefficient()
    {
        double stdDevX = standardDeviationX();
        double stdDevY = standardDeviationY();

        ValueHolder<Double> result = new ValueHolder<Double>(Double.valueOf(0));
        new Condition(() -> stdDevX == 0.0 || stdDevY == 0.0).ifFalse(() -> {
            result.set(Double.valueOf(covariance() / (stdDevX * stdDevY)));
        });

        this.coefficient = result.get();
        return this.coefficient;
    }

    /**
     * Returns the covariance.
     *
     * @return The covariance.
     */
    public Double covariance()
    {
        Integer n = collectionX.size();
        double meanX = meanX();
        double meanY = meanY();
        double cov = IntStream.range(0, n).parallel().mapToDouble(i -> (collectionX.get(i) - meanX) * (collectionY.get(i) - meanY)).sum();
        covariance = cov / (n - (nMinusOne ? 1 : 0));
        return covariance;
    }

    /**
     * Clears the collections of X and Y.
     */
    public void flush()
    {
        collectionX.clear();
        collectionY.clear();
    }

    /**
     * Clears the collection of X.
     */
    public void flushX()
    {
        collectionX.clear();
    }

    /**
     * Clears the collection of Y.
     */
    public void flushY()
    {
        collectionY.clear();
    }

    /**
     * Returns the mean of X.
     * It uses a parallel stream, converts each element to a double, and calculates the average. Returns 0.0 if the collection is empty.
     *
     * @return The mean of X.
     */
    public Double meanX()
    {
        meanX = collectionX.parallelStream().mapToDouble(eachElement -> eachElement.doubleValue()).average().orElse(0.0);
        return meanX;
    }

    /**
     * Returns the mean of Y.
     *
     * @return The mean of Y.
     */
    public Double meanY()
    {
        meanY = collectionY.parallelStream().mapToDouble(eachElement -> eachElement.doubleValue()).average().orElse(0.0);
        return meanY;
    }

    /**
     * Returns whether n-1 is used in calculations.
     *
     * @return true if the denominator is n-1, false if it is n.
     */
    public Boolean nMinusOne()
    {
        return nMinusOne;
    }

    /**
     * Toggles the denominator between n and n-1 for statistical calculations.
     *
     * @param aBoolean true to use n-1 as the denominator, false to use n.
     */
    public void nMinusOne(Boolean aBoolean)
    {
        this.nMinusOne = aBoolean;
    }

    /**
     * Returns the standard deviation of X.
     *
     * @return The standard deviation of X.
     */
    public Double standardDeviationX()
    {
        standardDeviationX = Math.sqrt(varianceX());
        return standardDeviationX;
    }

    /**
     * Returns the standard deviation of Y.
     *
     * @return The standard deviation of Y.
     */
    public Double standardDeviationY()
    {
        standardDeviationY = Math.sqrt(varianceY());
        return standardDeviationY;
    }

    /**
     * Returns the variance of X.
     *
     * @return The variance of X.
     */
    public Double varianceX()
    {
        double meanX = meanX();
        Integer n = collectionX.size();
        double sumX = collectionX.parallelStream().mapToDouble(eachElement -> Math.pow(eachElement - meanX, 2)).sum();
        varianceX = sumX / (n - (nMinusOne ? 1 : 0));
        return varianceX;
    }

    /**
     * Returns the variance of Y.
     *
     * @return The variance of Y.
     */
    public Double varianceY()
    {
        double meanY = meanY();
        Integer n = collectionY.size();
        double sumY = collectionY.parallelStream().mapToDouble(eachElement -> Math.pow(eachElement - meanY, 2)).sum();
        varianceY = sumY / (n - (nMinusOne ? 1 : 0));
        return varianceY;
    }

    /**
     * Returns the collection of X.
     *
     * @return The collection of X values.
     */
    public ArrayList<Double> xCollection()
    {
        return collectionX;
    }

    /**
     * Returns the collection of Y.
     *
     * @return The collection of Y values.
     */
    public ArrayList<Double> yCollection()
    {
        return collectionY;
    }

    /**
     * Updates the collection of X.
     *
     * @param xCollection The collection of X values.
     */
    public void xCollection(ArrayList<Double> xCollection)
    {
        this.collectionX = xCollection;
    }

    /**
     * Updates the collection of Y.
     *
     * @param yCollection The collection of Y values.
     */
    public void yCollection(ArrayList<Double> yCollection)
    {
        this.collectionY = yCollection;
    }

    /**
     * Returns the contents of the fields as a string.
     */
    public String toString()
    {
        return null;
    }
}

File: composeApp/src/commonMain/kotlin/app/typist/Correlation.kt
package app.typist

import kotlin.math.pow
import kotlin.math.sqrt

/**
 * A class for calculating the correlation coefficient for a given set of data.
 * This class is immutable; to perform a new calculation, create a new instance.
 *
 * @property xCollection The collection of X values.
 * @property yCollection The collection of Y values.
 * @property useNMinusOne A flag to indicate whether to use n-1 (for sample) or n (for population) in calculations.
 */
class Correlation(
    val xCollection: List<Double> = emptyList(),
    val yCollection: List<Double> = emptyList(),
    val useNMinusOne: Boolean = true
) {

    /** The mean of X, calculated once on first access. */
    val meanX: Double by lazy {
        xCollection.average()
    }

    /** The mean of Y, calculated once on first access. */
    val meanY: Double by lazy {
        yCollection.average()
    }

    /** The variance of X, calculated once on first access. */
    val varianceX: Double by lazy {
        val n = xCollection.size
        if (n < 2) return@lazy 0.0
        val sumX = xCollection.sumOf { (it - meanX).pow(2) }
        sumX / (n - if (useNMinusOne) 1 else 0)
    }

    /** The variance of Y, calculated once on first access. */
    val varianceY: Double by lazy {
        val n = yCollection.size
        if (n < 2) return@lazy 0.0
        val sumY = yCollection.sumOf { (it - meanY).pow(2) }
        sumY / (n - if (useNMinusOne) 1 else 0)
    }

    /** The standard deviation of X, calculated once on first access. */
    val standardDeviationX: Double by lazy {
        sqrt(varianceX)
    }

    /** The standard deviation of Y, calculated once on first access. */
    val standardDeviationY: Double by lazy {
        sqrt(varianceY)
    }

    /** The covariance of X and Y, calculated once on first access. */
    val covariance: Double by lazy {
        val n = xCollection.size
        if (n < 2) return@lazy 0.0
        val cov = (0 until n).sumOf { i ->
            (xCollection[i] - meanX) * (yCollection[i] - meanY)
        }
        cov / (n - if (useNMinusOne) 1 else 0)
    }

    /** The correlation coefficient, calculated once on first access. */
    val coefficient: Double by lazy {
        val denominator = standardDeviationX * standardDeviationY
        if (denominator == 0.0) 0.0 else covariance / denominator
    }

    /**
     * Returns the contents of the fields as a string.
     */
    override fun toString(): String {
        return "Correlation(xCollection=$xCollection, yCollection=$yCollection, useNMinusOne=$useNMinusOne)"
    }
}
